Runtime/LLMChatHistory.cs

```diff
@@ -0,0 +1,147 @@
+/// @file
+/// @brief File implementing the LLMChatHistory.
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using UnityEngine;
+
+namespace LLMUnity
+{
+    /// @ingroup llm
+    /// <summary>
+    /// Manages a single instance of a chat history.
+    /// </summary>
+    public class LLMChatHistory : MonoBehaviour
+    {
+        /// <summary> 
+        /// The name of the file where this chat history will be saved.
+        /// The file will be saved within the persistentDataPath directory (see https://docs.unity3d.com/ScriptReference/Application-persistentDataPath.html). 
+        /// </summary>
+        public string ChatHistoryFilename = string.Empty;
+
+        /// <summary>
+        /// If true, this component will automatically save a copy of its data to the filesystem with each update.
+        /// </summary>
+        public bool EnableAutoSave = true;
+
+        /// <summary>
+        /// The current chat history
+        /// </summary>
+        protected List<ChatMessage> _chatHistory;
+
+        /// <summary>
+        /// Ensures we're not trying to update the chat while saving or loading
+        /// </summary>
+        protected SemaphoreSlim chatLock = new SemaphoreSlim(1, 1);
+
+        /// <summary>
+        /// The Unity Awake function that initializes the state before the application starts.
+        /// </summary>
+        public async void Awake()
+        {
+            // If a filename has been provided for the chat history, attempt to load it
+            if (ChatHistoryFilename != string.Empty) {
+                await Load();
+            }
+            else {
+                _chatHistory = new List<ChatMessage>();
+            }
+        }
+
+        /// <summary>
+        /// Appends a new message to the end of this chat.
+        /// </summary>
+        public async Task AddMessage(string role, string content)
+        {
+            await WithChatLock(async () => {
+                await Task.Run(() => _chatHistory.Add(new ChatMessage { role = role, content = content }));
+            });
+
+            if (EnableAutoSave) {
+                _ = Save();
+            }
+        }
+
+        public List<ChatMessage> GetChatMessages() {
+            return new List<ChatMessage>(_chatHistory);
+        }
+
+        /// <summary>
+        /// Saves the chat history to the file system
+        /// </summary>
+        public async Task Save()
+        {
+            // If no filename has been provided, create one
+            if (ChatHistoryFilename == string.Empty) {
+                ChatHistoryFilename = $"chat_{Guid.NewGuid()}";
```

Comment:
The problem with creating a new filename is that the file won't be usable since it will not be persisted anywhere.
If someones wants to use the chat history, they would need to get and store the ChatHistoryFilename somewhere at runtime.
Is there some way to automatically create a new LLMChatHistory GameObject for a LLMCharacter when added to the scene?
This way we could have similar behavior as before, the use would need to specify the chat filename if they want to autosave

Reply:
So in the LLMCharacter InitHistory() we already create and add a chat history component if ther isn't one. I'll just have it set the filename there to match the cache filename.

Reply:
Hmm, I guess my thought was that we need to save it somewhere. If the developer doesn't provide us with any name (either from the character or from the ChatHistoryFilename) then we've got to do something. We could just require them to provide a name instead?